packages/language-service/src/lib-new/lang-service-context.ts

```diff
@@ -0,0 +1,203 @@
+import { TextDocument } from 'vscode-css-languageservice';
+import { CSSResolve, Diagnostics, Stylable, StylableMeta } from '@stylable/core';
+import type { StylableFile } from '../lib/service';
+import { getAstNodeAt } from './ast-from-position';
+import { parseForEditing, ParseReport } from './edit-time-parser';
+import {
+    parseSelectorWithCache,
+    StylableProcessor,
+    StylableTransformer,
+} from '@stylable/core/dist/index-internal';
+import { URI } from 'vscode-uri';
+import {
+    ImmutableSelector,
+    ImmutableSelectorList,
+    ImmutableSelectorNode,
+    SelectorList,
+    stringifySelectorAst,
+    walk as walkSelector,
+} from '@tokey/css-selector-parser';
+import * as postcss from 'postcss';
+
+type ResolveChainItem = {
+    [Type in ImmutableSelectorNode['type']]?: Extract<ImmutableSelectorNode, { type: Type }>[];
+} & {
+    resolved?: CSSResolve[];
+    resolvedNode?: ImmutableSelectorNode;
+};
+
+export class LangServiceContext {
+    public meta: StylableMeta;
+    public errorNodes: Map<any, ParseReport[]>;
+    public ambiguousNodes: Map<any, ParseReport[]>;
+    public location: ReturnType<typeof getAstNodeAt>;
+    public document: TextDocument;
+    constructor(public stylable: Stylable, private fileData: StylableFile, private offset: number) {
+        const parseResult = parseForEditing(fileData.content, {
+            from: fileData.path,
+        });
+        this.errorNodes = parseResult.errorNodes;
+        this.ambiguousNodes = parseResult.ambiguousNodes;
+        this.location = getAstNodeAt(parseResult, offset);
+        this.meta = new StylableProcessor(
+            new Diagnostics(),
+            this.stylable.resolveNamespace
+        ).process(parseResult.ast);
+        this.document = TextDocument.create(
+            URI.file(this.meta.source).toString(),
+            'stylable',
+            this.fileData.stat.mtime.getTime(),
+            this.fileData.content
+        );
+    }
+    public getPosition(offset: number = this.offset) {
+        return this.document.positionAt(offset);
+    }
+    public isInRoot() {
+        return this.location.base.node.type === 'root';
+    }
+    public isInSelector() {
+        return !!this.location.selector;
+    }
+    public getSelectorContext() {
+        const nestedSelectors = this.collectSelector();
+        // collect the CSS resolve chain for each selector node
+        const selectorAstResolveMap = new Map<ImmutableSelectorNode, CSSResolve[]>();
+        let currentAnchor = undefined;
+        for (const selector of nestedSelectors) {
```

Comment:
do we want to run on each selector in the list separately?

Reply:
This will part of handling intersection types and nesting. 
At the moment we have almost no cases where we get multiple selectors here, I think only for `@st-scope` with multiple selectors, and the transformer is buggy for that case anyhow.